src/Tasks/LargeFileUploadTask.php

```diff
@@ -0,0 +1,345 @@
+    public static function createUploadSession(RequestAdapter $adapter, LargeFileUploadCreateUploadSessionBody $uploadSessionBody, string $url): Promise {
+        $requestInformation = new RequestInformation();
+        $baseUrl = rtrim($adapter->getBaseUrl(), '/');
+        $path = ltrim($url, '/');
+
+        $newUrl = "{$baseUrl}/$path";
+        $requestInformation->setUri($newUrl);
+        $requestInformation->httpMethod = HttpMethod::POST;
+        $requestInformation->setContentFromParsable($adapter, 'application/json', $uploadSessionBody);
+        return $adapter->sendAsync($requestInformation, [LargeFileTaskUploadSession::class, 'createFromDiscriminatorValue']);
+    }
+    /**
+     * @return RequestAdapter
+     */
+    public function getAdapter(): RequestAdapter {
+        return $this->adapter;
+    }
+
+    /**
+     * @return int
+     */
+    public function getChunks(): int {
+        return $this->chunks;
+    }
+
+    /**
+     * @return int
+     */
+    public function getUploadedChunks(): int {
+        return $this->uploadedChunks;
+    }
+
+    /**
+     * @param Parsable|null $uploadSession
+     * @throws Exception
+     */
+    public function uploadSessionExpired(?Parsable $uploadSession): bool {
+        $now = new DateTime((new DateTime('now'))->format(DateTimeInterface::ATOM));
+
+        $validatedValue = $this->checkValueExists($uploadSession ?? $this->uploadSession, 'getExpirationDateTime', ['ExpirationDateTime', 'expirationDateTime']);
+        if (!$validatedValue[0]) {
+            throw new Exception('The upload session does not contain an expiry datetime.');
+        }
+        $expiry = $validatedValue[1];
+
+        if ($expiry === null){
+            throw new InvalidArgumentException('The upload session does not contain a valid expiry date.');
+        }
+        $then = new DateTime($expiry->format(DateTimeInterface::ATOM));
+        $interval = $now->diff($then);
+
+        if ($interval->invert !== 0){
+            return true;
+        }
+        return false;
+    }
+    /**
+     * @throws Exception
+     */
+    public function upload(?callable $afterChunkUpload = null): void {
+
+        if ($this->uploadSessionExpired($this->uploadSession)){
+            throw new RuntimeException('The upload session is expired.');
+        }
+        $q = new SplQueue();
+
+        $start = 0;
+        $session = $this->nextChunk($this->stream, $start,max(0, min($this->maxChunkSize - 1,  $this->fileSize - 1)));
+        $q->enqueue($session);
+
+        while(!$q->isEmpty()){
+            /** @var Promise $front */
+            $front = $q->dequeue();
+
+            $front->then(function (LargeFileTaskUploadSession $session) use (&$q, $afterChunkUpload){
+                $nextRange = $session->getNextExpectedRanges();
+                $this->uploaded = (int)explode('-', $nextRange[0] ?? ($this->fileSize.'-'))[0];
+                if (empty($nextRange)) {
+                    echo "Upload finished!!!!\n";
```

Comment:
Left by mistake?
Also, when the final byte range is uploaded & `deferCommit` was false, the final object returned is a `DriveItem`. https://learn.microsoft.com/en-us/graph/api/driveitem-createuploadsession?view=graph-rest-1.0#completing-a-file

Does this work? considering the callback expects a `LargeFileTaskUploadSession`?

Could we add a `complete()` method which would complete the upload session in case `deferCommit` was true?

Reply:
Removing the print. 
For the `complete` I think that might be left to the user to do when they choose to `deferCommit`. This is done by returning the `UploadSession` with the `uploadUrl` when you upload the last chunk, which the user can use to send a post request to complete the upload.

Reply:
This is a decision partly motivated by the differences in the Print, OneDrive and Outlook APIs.

Reply:
Make sense, might be good to align on this design-wise. Previous design seemed to have a result object.